File: python/leetcode/balanced-btree.py
"""
Given a binary tree, determine if it is height-balanced. For this problem, 
a height-balanced binary tree is defined as: a binary tree in which the depth 
of the two subtrees of every node never differ by more than 1.
"""


class Solution:
  def isBalanced(self, root):
    if not root:
      return True

    return self.isBalanced(root.left) and self.isBalanced(root.right) \
      and abs(self.treeDepth(root.left) - self.treeDepth(root.right)) <= 1

  def treeDepth(self, node):
    if not node:
      return 0

    return 1 + max(self.treeDepth(node.left), self.treeDepth(node.right))

File: java/leetcode/balanced-btree.java
/*
Given a binary tree, determine if it is height-balanced. For this problem, 
a height-balanced binary tree is defined as: a binary tree in which the depth 
of the two subtrees of every node never differ by more than 1.
*/

class Solution {
  public boolean isBalanced(TreeNode root) {
    if (root == null) {
      return true;
    }

    return isBalanced(root.left) && isBalanced(root.right)
        && Math.abs(treeDepth(root.left) - treeDepth(root.right)) <= 1;
  }

  public int treeDepth(TreeNode node) {
    if (node == null) {
      return 0;
    }

    return 1 + Math.max(treeDepth(node.left), treeDepth(node.right));
  }
}